08-read-write-plot.Rmd

```diff
@@ -384,329 +112,593 @@ Executing commands such as `sf::read_sf()` (the main function we use for loading
+
+\index{raster!data input}
+Similar to vector data, raster data comes in many file formats with some of them supporting multilayer files.
+**terra**'s `rast()` command reads in a single layer when a file with just one layer is provided.
+
+```{r 07-read-write-plot-24, message=FALSE}
+raster_filepath = system.file("raster/srtm.tif", package = "spDataLarge")
+single_layer = rast(raster_filepath)
+```
+
+It also works in case you want to read a multilayer file.
+
+```{r 07-read-write-plot-25}
+multilayer_filepath = system.file("raster/landsat.tif", package = "spDataLarge")
+multilayer_rast = rast(multilayer_filepath)
+```
+
+\index{vsicurl}
+\index{GDAL}
+\index{COG}
+All of the previous examples read spatial information from files stored on your hard drive. 
+However, GDAL also allows reading data directly from online resources, such as HTTP/HTTPS/FTP web resources.
+The only thing we need to do is to add a `/vsicurl/` prefix before the path to the file.
+Let's try it by connecting to the global monthly snow probability at 500 m resolution for the period 2000-2012 [@hengl_t_2021_5774954].
+Snow probability for December is stored as a Cloud Optimized GeoTIFF (COG) file (see Section \@ref(file-formats)) at [zenodo.org](https://zenodo.org/record/5774954/files/clm_snow.prob_esacci.dec_p.90_500m_s0..0cm_2000..2012_v2.0.tif).
+To read an online file, we just need to provide its URL together with the `/vsicurl/` prefix.
+
+```{r}
+myurl = "/vsicurl/https://zenodo.org/record/5774954/files/clm_snow.prob_esacci.dec_p.90_500m_s0..0cm_2000..2012_v2.0.tif"
+snow = rast(myurl)
+snow
+```
+
+Due to the fact that the input data is COG, we are actually not reading this file to our RAM, but rather creating a connection to it without obtaining any values.
+Its values will be read if we apply any value-based operation (e.g., `crop()` or `extract()`).
+This allows us also to just read a tiny portion of the data without downloading the entire file.
+For example, we can get the snow probability for December in Reykjavik (70%) by specifying its coordinates and applying the `extract()` function:
+
+```{r}
+rey = data.frame(lon = -21.94, lat = 64.15)
+snow_rey = extract(snow, rey)
+snow_rey
+```
+
+This way, we just downloaded a single value instead of the whole, large GeoTIFF file.
+
+The above example just shows one simple (but useful) case, but there is more to explore.
+The `/vsicurl/` prefix also works not only for raster but also for vector file formats.
+It allows reading vectors directly from online storage with `read_sf()` just by adding the prefix before the vector file URL.
+
+Importantly, `/vsicurl/` is not the only prefix provided by GDAL -- many more exist, such as `/vsizip/` to read spatial files from ZIP archives without decompressing them beforehand or `/vsis3/` for on-the-fly reading files available in AWS S3 buckets.
+You can learn more about it at https://gdal.org/user/virtual_file_systems.html.
+
+<!-- ### Databases -->
+<!-- jn:toDo-->
```

Comment:
It is already in chapter 10: https://r.geocompx.org/gis#postgis. I think that for some time I considered moving it to I/O, but at the end it can stay where it is now.

Reply:
OK, we should add a cross-reference I think.

Reply:
> It is already in chapter 10: https://r.geocompx.org/gis#postgis. I think that for some time I considered moving it to I/O, but at the end it can stay where it is now.

Done: https://github.com/geocompx/geocompr/pull/1032/commits/1a798eb8528baec4ce458f6e3bc121bd552f9f1f

Reply:
xref added.